[tests] sync_with_ping should assert that ping hasn't timed out

jnewbery opened the pull request:
sync_with_ping() currently returns false if the timeout expires, and it is
the caller's responsibility to fail the test. However, none of the tests
currently assert on sync_with_ping()'s return code. This commit adds an
assert to sync_with_ping so the test will fail if the timeout expires.

This commit also removes all the duplicate implementations of
sync_with_ping() from the individual tests.

~Only the second commit here needs to be reviewed. The first commit https://github.com/bitcoin/bitcoin/commit/159fe88abfadf67410578c145e631819cf50b660 is #10109.~ __EDIT: #10109 is now merged.__

__EDIT: The first commit in this PR ("Add send_await_disconnect() method to p2p-compactblocks.py") is a test case bug fix that needs to be merged before merging the sync_with_ping() change__

Should fix the intermittent failure reported here: https://github.com/bitcoin/bitcoin/pull/10073#issuecomment-289466653

sync'ing and pinging @sdaftuar

jnewbery committed:
[tests] sync_with_ping should assert that ping hasn't timed out

sync_with_ping currently returns false if the timeout expires, and it is
the caller's responsibility to fail the test. However, none of the tests
currently assert on sync_with_ping()'s return code. This commit adds an
assert to sync_with_ping so the test will fail if the timeout expires.

This commit also removes all the duplicate implementations of
sync_with_ping() from the individual tests.

sdaftuar commented:
I think travis may have caught a bug in the p2p-compactblocks.py test, which is now throwing an error that the sync with ping failed?

jnewbery committed:
Add send_await_disconnect() method to p2p-compactblocks.py

p2p-compactblocks was incorrectly using sync_with_ping() when sending in
invalid block. The node would disconnect us and never respond to the
ping, so the sync_with_ping would just time out after 30 seconds and
continue with the test.

This commit adds a send_await_disconnect() method that sends the
message, and then waits for the node to disconnect us. In this commit
I've added the method to p2p-compactblocks.py, but a future commit could
move it to mininode since it could be useful more generally.

This commit reduces the p2p-compactblock runtime by 30 seconds.

jnewbery commented:
Oh good. We've caught our first fish.

I've pushed a commit that changes p2p-compactblock to use a new method send_await_disconnect(), which awaits a disconnect rather than a pong. 

The commit adds the method to p2p-compactblocks.py, but a future commit could move it to mininode since it could be useful more generally. The commit reduces the p2p-compactblock runtime by 30 seconds.

I've added the new commit _before_ the commit that changes sync_with_ping() so git_bisect doesn't break.

laanwj commented:
utACK 6a18bb9

sdaftuar commented:
utACK 6a18bb9
